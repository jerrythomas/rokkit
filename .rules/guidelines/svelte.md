# Guidelines

- Use Svelte 5 runes mode for code
- Use SvelteKit for app
- Use UnoCSS for styling
- Use vitest for unit testing
- Use Playwright for e2e testing
- E2E tests should follow the page object pattern

## Svelte 5
{{[Svelte & Svelte Kit Distilled](https://svelte-llm.khromov.se/svelte-complete-distilled)}}
{{[svelte 5 documentation](https://svelte.dev/docs/svelte/llms.txt)}}

## Snippets

### Runes

Runes are special functions that replace reactivity declarations from Svelte 4.

```svelte
<script>
  // State
  let count = $state(0);

  // Derived values
  let doubled = $derived(count * 2);

  // Props
  let {
    name = 'world',
    value = $bindable(null)
  } = $props();

  // Effects
  $effect(() => {
    console.log(`Count changed to ${count}`);
  });
</script>
```

### Defining and passing snippets to components:

```svelte
<script>
  let items = ['one','two', 'three']
</script>
<!-- a header snippet -->
{#snippet header()}
  <h1>Custom header</h1>
{/snippet}

<!-- a snippet with parameters -->
{#snippet item(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {items} {header} {item} />
```

### Using snippets in a component:

```svelte
<script>
  let { header, item } = $props();
</script>

<div class="header">
  {@render header?.()}
</div>

{#each items as currentItem}
  {@render item(currentItem, selectedItem === currentItem)}
{/each}
```

To pass a snippet to another component:

#### ChildComponent.svelte

```svelte
<script>
  let { header } = $props();
</script>

{#if header}
  {@render header()}
{/if}
```

#### Snippets can be passed as properties to child component

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
{#snippet header()}
  <h1>Custom Header</h1>
{/snippet}

<ChildComponent {header} />
```

### Snippets can be enclosed within the child component.

```svelte
<script>
   import  ChildComponent from './ChildComponent.svelte'
</script>
<ChildComponent>
  {#snippet header()}
    <h1>Custom Header</h1>
  {/snippet}
</ChildComponent>
```
Both of these options work in a similar way

### Important: What NOT to do

Do not confuse snippets with React-style JSX or string templates:

```svelte
<!-- ❌ INCORRECT - Using arrow functions returning strings -->
<List
  header={() => `<h1>My List</h1>`}
  item={({ item }) => `<div>${item.name}</div>`}
/>

<!-- ✅ CORRECT - Using properly defined snippets -->
<!-- snippets are defined in the html section not in scripts block -->
{#snippet header()}
  <h1>My List</h1>
{/snippet}

{#snippet itemTemplate(item, selected)}
  <div class:selected={selected}>
    {item.name}
  </div>
{/snippet}

<List {header} item={itemTemplate} />
```

## Component Design Guidelines

- should be modular and data driven
- should be accessibile
- Reduced motion should be considered

## Project Structure

If the source code uses runes or the test uses runes the file extension should be `.svelte.js` or `.spec.svelte.js` respectively.

```
project/
│
├── src/
│   ├── index.js
│   ├── Component.svelte
│   └── utils.svelte.js
│
├── spec/
│   ├── index.spec.js
│   ├── Component.spec.js
│   └── utils.spec.svelte.js
├── e2e/
│   ├── home.e2e.js
│   └── login.e2e.js
└── README.md
```
