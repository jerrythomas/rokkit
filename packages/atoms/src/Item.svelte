<script>
	import Icon from './Icon.svelte'
	import { FieldMapper } from '@rokkit/core'

	/**
	 * @type {Object} Props
	 * @property {string|Object} [value]
	 * @property {FieldMapper}   [mapping]
	 */
	let { value, mapping = new FieldMapper() } = $props()
</script>

{#if mapping.hasIcon(value)}
	<Icon name={mapping.getIcon(value)} label={mapping.getLabel(value)} />
{:else if mapping.hasImage(value)}
	<img src={mapping.getImage(value)} alt={mapping.getLabel(value)} />
{/if}
<p>{mapping.getText(value)}</p>
