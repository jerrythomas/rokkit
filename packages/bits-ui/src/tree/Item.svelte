<script>
	/**
	 * @typedef {Object} Props
	 * @property {string} [class]
	 */

	/** @type {Props & { [key: string]: any }} */
	let { class: className = '', children, ...rest } = $props()
</script>

<tree-item {...rest} class={className} data-tree-item>
	{@render children?.()}
</tree-item>
