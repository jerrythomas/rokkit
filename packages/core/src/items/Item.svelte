<script>
	import { defaultFields } from '../constants'
	import Icon from '../Icon.svelte'

	export let value
	export let fields = defaultFields

	$: isObject = typeof value == 'object'
	$: text = isObject ? value[fields.text] : value
</script>

{#if isObject}
	{#if value[fields.icon]}
		{@const iconName =
			typeof value[fields.icon] == 'object'
				? value[fields.icon][value[fields.state]]
				: value[fields.icon]}
		<Icon name={iconName} />
	{:else if value[fields.image]}
		<img
			class="h-4 w-4 object-cover"
			alt={value[fields.text]}
			src={value[fields.image]}
		/>
	{/if}
{/if}
{#if text}
	<p class="w-full flex">{text}</p>
{/if}
