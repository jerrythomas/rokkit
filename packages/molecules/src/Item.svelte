<script>
	import { defaultFields, getIcon, getText } from '@rokkit/core'
	import { Icon } from '@rokkit/atoms'

	export let value
	export let fields = defaultFields

	$: isObject = typeof (value ?? '-') == 'object'
	$: text = getText(value, fields)
</script>

{#if isObject}
	{#if value[fields.icon]}
		{@const iconName = getIcon(value, fields)}
		<Icon name={iconName} />
	{:else if value[fields.image]}
		<img class="min-h-4 min-w-4 object-cover" alt={value[fields.text]} src={value[fields.image]} />
	{/if}
{/if}
{#if text}
	<p>{text}</p>
{/if}
