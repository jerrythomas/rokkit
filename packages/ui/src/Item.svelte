<script>
	import Icon from './Icon.svelte'
	import { FieldMapper } from '@rokkit/core'

	/**
	 * @typedef {Object} Props
	 * @property {string|Object} [value]
	 * @property {FieldMapper}   [mapping]
	 */

	/** @type {Props} */
	let { value, mapping = new FieldMapper() } = $props()
	let content = $derived(mapping.getText(value))
	let ariaLabel = $derived(mapping.getLabel(value) ?? content)
</script>

{#if mapping.hasIcon(value)}
	<Icon name={mapping.getIcon(value)} label={ariaLabel} />
{:else if mapping.hasImage(value)}
	<img src={mapping.getImage(value)} alt={ariaLabel} />
{/if}
{#if content}
	<p>{content}</p>
{/if}
