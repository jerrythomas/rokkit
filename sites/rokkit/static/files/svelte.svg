<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
<path d="M18.2301 11.1217C21.1087 6.47773 15.3383 1.6935 11.2372 4.37297L7.337 6.85957C5.44834 8.08961 4.62832 10.453 5.78004 12.8817C2.90038 17.2501 8.50365 22.3655 12.7729 19.6332L16.6732 17.1466C18.7952 15.7494 19.2821 13.3398 18.2301 11.1217V11.1217ZM8.46536 8.46761L12.3599 5.98242C14.9184 4.45077 17.8644 7.12896 16.8023 9.62435C16.1715 9.12318 15.4685 8.81503 14.9345 8.67627C15.1375 7.87006 14.2016 7.27202 13.5507 7.68841L9.64763 10.1651C8.68258 10.8321 9.71148 12.204 10.6653 11.6284L12.1527 10.6803C15.4905 8.79676 18.6495 13.3192 15.5448 15.5357L11.6502 18.0209C9.30336 19.5215 6.10922 17.1175 7.20785 14.379C7.79145 14.8207 8.41131 15.1456 9.07564 15.3271C8.92896 16.2531 9.86333 16.7033 10.4694 16.3234L14.3626 13.8382C15.3233 13.1178 14.292 11.814 13.3464 12.375L11.8533 13.3231C8.55362 15.2514 5.32106 10.6843 8.46536 8.46761Z" fill="black"/>
</svg>
