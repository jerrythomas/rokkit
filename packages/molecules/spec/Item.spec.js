import { describe, expect, it, beforeEach } from 'vitest'
import { cleanup, render } from '@testing-library/svelte'
import { tick } from 'svelte'
import Item from '../src/Item.svelte'

describe('Item.svelte', () => {
	beforeEach(() => cleanup())

	it('should render default', async () => {
		const { container, component } = render(Item, { value: null })
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()

		let text = container.querySelector('p')
		expect(text).toBeFalsy()

		// handle value change
		component.$set({ value: 'hello' })
		await tick()
		text = container.querySelector('p')
		expect(text).toBeTruthy()
		expect(text.textContent).toBe('hello')
	})

	it('should render null', () => {
		const { container } = render(Item, { value: null })
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render with value', () => {
		const { container } = render(Item, { value: 'text' })
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render object', () => {
		const { container } = render(Item, { value: { text: 'hello' } })
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render icon', () => {
		const { container } = render(Item, {
			value: { text: 'hello', icon: 'info' }
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})
	it('should render image', () => {
		const { container } = render(Item, {
			value: { text: 'hello', image: 'https://example.com/img.png' }
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})
	it('should render image and icon', () => {
		const { container } = render(Item, {
			value: {
				text: 'hello',
				image: 'https://example.com/img.png',
				icon: 'info'
			}
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})
	it('should render icon based on state', () => {
		const { container } = render(Item, {
			value: {
				text: 'hello',
				icon: { on: 'info', off: 'info-off' },
				state: 'on'
			}
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})
	it('should render using field mapping', () => {
		const { container } = render(Item, {
			value: {
				alt: 'hello',
				profile: 'https://example.com/img.png',
				ico: 'info'
			},
			fields: { text: 'alt', image: 'profile', icon: 'ico' }
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	it('should render empty when field mapping is invalid', () => {
		const { container } = render(Item, {
			value: {
				alt: 'hello'
			}
		})
		expect(container).toBeTruthy()
		expect(container).toMatchSnapshot()
	})

	// it('should update when value changes', async () => {
	// 	const { container, component } = render(Item, {
	// 		value: {
	// 			text: 'hello'
	// 		}
	// 	})
	// 	expect(container).toBeTruthy()
	// 	const text = container.querySelector('p')
	// 	expect(container).toMatchSnapshot()
	// 	expect(text.textContent).toBe('hello')

	// 	component.$set({ value: { text: 'world' } })
	// 	await tick()
	// 	expect(text.textContent).toBe('world')
	// })
})
