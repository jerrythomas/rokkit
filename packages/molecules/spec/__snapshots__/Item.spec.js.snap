// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Item.svelte > should render default 1`] = `
<div>
  <!---->
  <!---->
   
  <!---->
  
</div>
`;

exports[`Item.svelte > should render empty when field mapping is invalid 1`] = `
<div>
  <!---->
  <!---->
  <!---->
  
  
  <!---->
   
  <!---->
  
</div>
`;

exports[`Item.svelte > should render icon 1`] = `
<div>
  <!---->
  <!---->
  <rkt-icon
    aria-label="info"
    class="flex flex-shrink-0 items-center justify-center "
    role="img"
    tabindex="-1"
  >
    <i
      aria-hidden="true"
      class="info"
    />
  </rkt-icon>
  
  <!---->
   
  <p>
    hello
  </p>
  <!---->
  
</div>
`;

exports[`Item.svelte > should render icon based on state 1`] = `
<div>
  <!---->
  <!---->
  <rkt-icon
    aria-label="info"
    class="flex flex-shrink-0 items-center justify-center "
    role="img"
    tabindex="-1"
  >
    <i
      aria-hidden="true"
      class="info"
    />
  </rkt-icon>
  
  <!---->
   
  <p>
    hello
  </p>
  <!---->
  
</div>
`;

exports[`Item.svelte > should render image 1`] = `
<div>
  <!---->
  <!---->
  <!---->
  <img
    alt="hello"
    class="min-h-4 min-w-4 object-cover"
    src="https://example.com/img.png"
  />
  
  
  <!---->
   
  <p>
    hello
  </p>
  <!---->
  
</div>
`;

exports[`Item.svelte > should render image and icon 1`] = `
<div>
  <!---->
  <!---->
  <rkt-icon
    aria-label="info"
    class="flex flex-shrink-0 items-center justify-center "
    role="img"
    tabindex="-1"
  >
    <i
      aria-hidden="true"
      class="info"
    />
  </rkt-icon>
  
  <!---->
   
  <p>
    hello
  </p>
  <!---->
  
</div>
`;

exports[`Item.svelte > should render null 1`] = `
<div>
  <!---->
  <!---->
   
  <!---->
  
</div>
`;

exports[`Item.svelte > should render object 1`] = `
<div>
  <!---->
  <!---->
  <!---->
  
  
  <!---->
   
  <p>
    hello
  </p>
  <!---->
  
</div>
`;

exports[`Item.svelte > should render using field mapping 1`] = `
<div>
  <!---->
  <!---->
  <rkt-icon
    aria-label="info"
    class="flex flex-shrink-0 items-center justify-center "
    role="img"
    tabindex="-1"
  >
    <i
      aria-hidden="true"
      class="info"
    />
  </rkt-icon>
  
  <!---->
   
  <p>
    hello
  </p>
  <!---->
  
</div>
`;

exports[`Item.svelte > should render with value 1`] = `
<div>
  <!---->
  <!---->
   
  <p>
    text
  </p>
  <!---->
  
</div>
`;
